File: Python/src/graphs/undirected/graph_edges.py
from __future__ import annotations

from collections.abc import Callable, Iterable
from typing import Generic, TypeVar

from .abstract_graph import AbstractGraph
from .edge import Edge
from .edge_graph import EdgeGraph

V = TypeVar("V")
E = TypeVar("E")


class GraphEdges(Generic[V, E], AbstractGraph[V, Edge[V, E]], EdgeGraph[V, E]):
    """
    A graph whose adjacency type is the Edge itself, so edges can carry attributes.

    NOTE the EdgeGraph base was missing until now, though the Java has
    ``implements EdgeGraph<V, E>``. Two things followed from that: EdgeGraph was
    an interface with no implementation anywhere in the Python tree, and
    ``add_edge_vertices`` -- the Java's ``addEdge(from, to, attribute)`` -- could
    not be called on the only class that should have had it.
    """

    __slots__ = ()

    def edges(self) -> Iterable[Edge[V, E]]:
        """
        NOTE an edge sits in the adjacency bag of BOTH its endpoints, so gathering
        every bag would report each edge twice. An edge is therefore collected only
        from the bag of the vertex ``Edge.get`` returns, which is one endpoint and
        not the other. A self-loop occupies one bag once, and is reported once.

        :return: every edge, each appearing exactly once.
        """
        result: list[Edge[V, E]] = []
        for vertex, bag in self._adjacent_edges.items():
            for e in bag:
                if e.get() == vertex:
                    result.append(e)
        return result

    def add_edge(
        self, edge: Edge[V, E], predicate: Callable[[Edge[V, E]], bool] | None = None
    ) -> None:
        """
        Add an edge to both of its vertices, because this graph is undirected and
        an edge is incident on each of its endpoints alike.

        Both bags, so that ``adjacent(v)`` reports the edges AT v rather than the
        edges that happen to have been written with v first -- which is what an
        algorithm walking the graph by adjacency needs. ``edges`` still reports
        each edge once; see there for how.

        :param edge: the edge to add.
        :param predicate: if given, the edge is added only when this accepts it.
        """
        if predicate is None or predicate(edge):
            v = edge.get()
            w = edge.get_other(v)
            self.get_adjacency_bag(v).add(edge)
            # A self-loop is incident on one vertex, so it belongs in one bag once.
            if v == w:
                return
            self.get_adjacency_bag(w).add(edge)

    def __str__(self) -> str:
        return str(self._adjacent_edges)

File: Java/src/main/java/com/phasmidsoftware/dsaipg/graphs/undirected/Graph_Edges.java
package com.phasmidsoftware.dsaipg.graphs.undirected;

import com.phasmidsoftware.dsaipg.adt.bqs.Bag;
import com.phasmidsoftware.dsaipg.adt.bqs.Bag_Array;
import com.phasmidsoftware.dsaipg.util.iteration.SizedIterable;

import java.util.Map;
import java.util.Objects;
import java.util.function.Predicate;

/**
 * Represents a graph structure where vertices are connected by edges, with support for edge-specific attributes.
 * This class extends the {@code AbstractGraph} class for generic graph behavior and implements the
 * {@code EdgeGraph} interface for edge-based operations.
 *
 * @param <V> the vertex type.
 * @param <E> the edge attribute type.
 */
public class Graph_Edges<V, E> extends AbstractGraph<V, Edge<V, E>> implements EdgeGraph<V, E> {

    /**
     * Retrieves all edges in the graph as a collection, each appearing exactly once.
     * <p>
     * NOTE an edge sits in the adjacency bag of BOTH its endpoints, so gathering
     * every bag would report each edge twice. An edge is therefore collected only
     * from the bag of the vertex that {@link Edge#get} returns, which is one
     * endpoint and not the other. A self-loop occupies one bag once, and is
     * likewise reported once.
     *
     * @return a SizedIterable containing all edges present in the graph.
     */
    public SizedIterable<Edge<V, E>> edges() {
        Bag<Edge<V, E>> result = new Bag_Array<>();
        for (Map.Entry<V, Bag<Edge<V, E>>> entry : adjacentEdges.entrySet())
            for (Edge<V, E> e : entry.getValue())
                if (Objects.equals(e.get(), entry.getKey()))
                    result.add(e);
        return result;
    }

    /**
     * Adds an edge to the graph if it satisfies the given predicate condition.
     * The edge is added to the adjacency bag of BOTH its vertices, because this
     * graph is undirected and an edge is incident on each of its endpoints alike.
     * <p>
     * NOTE both bags, so that {@code adjacent(v)} reports the edges AT v rather
     * than the edges that happen to have been written with v first -- which is what
     * an algorithm walking the graph by adjacency needs. {@link #edges} reports
     * each edge once all the same; see there for how.
     * {@link Graph_Simple#addEdge} records both directions too.
     *
     * @param edge      the edge to be added, defined by its two vertices and optional attributes.
     * @param predicate a condition that determines whether the edge should be added to the graph.
     */
    public void addEdge(Edge<V, E> edge, Predicate<Edge<V, E>> predicate) {
        if (predicate.test(edge)) {
            V v = edge.get(), w = edge.getOther(v);
            getAdjacencyBag(v).add(edge);
            // A self-loop is incident on one vertex, so it belongs in one bag once.
            if (Objects.equals(v, w)) return;
            getAdjacencyBag(w).add(edge);
        }
    }

    /**
     * Adds an edge to the graph using the specified vertices, attribute, and predicate.
     * The edge is only added if the given predicate evaluates to true.
     *
     * @param from      the starting vertex of the edge.
     * @param to        the ending vertex of the edge.
     * @param attribute the attribute associated with the edge.
     * @param predicate a condition to determine if the edge should be added.
     */
    public void addEdge(V from, V to, E attribute, Predicate<Edge<V, E>> predicate) {
        addEdge(new Edge<>(from, to, attribute), predicate);
    }

    @Override
    public String toString() {
        return adjacentEdges.toString();
    }

}
